packages/libs/client/src/v2/client/utils/request.ts

```diff
@@ -0,0 +1,139 @@
+  },
+  method: 'POST' as const,
+  body: JSON.stringify(body),
+});
+
+export async function parseResponse<T extends object | string>(
+  response: Response,
+): Promise<T> {
+  if (response.ok) {
+    return response.json();
+  } else {
+    try {
+      const textResponse: string = await response.text();
+      return Promise.reject(new Error(textResponse));
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  }
+}
+
+export function getUrl(
+  host: string,
+  endpoint: string,
+  params?: object,
+): string {
+  const urlStr = `${host}${
+    endpoint.startsWith('/') ? endpoint : `/${endpoint}`
+  }`;
+  const url = new URL(urlStr);
+
+  if (params !== undefined) {
+    Object.entries(params).forEach(([key, value]) => {
+      if (value !== undefined && value !== null) {
+        url.searchParams.append(key, value.toString());
+      }
+    });
+  }
+
+  return url.toString();
+}
+
+export const kadenaHostGenerator = (networkId: string, chainId: string) => {
+  switch (networkId) {
+    case 'mainnet01':
+      return `https://api.chainweb.com/chainweb/0.0/${networkId}/chain/${chainId}/pact`;
+    case 'testnet04':
+      return `https://api.testnet.chainweb.com/chainweb/0.0/${networkId}/chain/${chainId}/pact`;
+    default:
+      throw new Error(`UNKNOWN_NETWORK_ID: ${networkId}`);
+  }
+};
+
+export interface ICommandRequest {
+  cmd: string;
+  hash: string;
+  sigs: string[];
+}
+
+export type ICommandRequestWithoutHash = Omit<ICommandRequest, 'hash'>;
+
+export interface INetworkOptions {
+  networkId: string;
+  chainId: string;
+}
+
+export interface IPollOptions {
+  onPoll?: (id: string) => void;
+  timeout?: number;
+  interval?: number;
+}
+
+export interface IExtPromise<T> {
+  promise: Promise<T>;
+  resolve: (result: T) => void;
+  reject: (err: any) => void;
+  fulfilled: boolean;
+  data: T | undefined;
+}
+
+export const getPromise = <T>(): IExtPromise<T> => {
+  let resolve: (value: T) => void = () => {};
+  let reject: (value: any) => void = () => {};
+  let fulfilled = false;
+  let result: T | undefined;
+
+  const promise = new Promise<T>((_resolve, _reject) => {
+    resolve = (data: T) => {
+      result = data;
+      fulfilled = true;
+      _resolve(data);
+    };
+    reject = (err) => {
+      fulfilled = true;
+      _reject(err);
+    };
+  });
+
+  return {
+    promise,
+    resolve,
+    reject,
+    get fulfilled() {
+      return fulfilled;
+    },
+    get data() {
+      return result;
+    },
+  };
+};
+
+export const mergeAll = <T extends object>(results: Array<T>): T =>
+  results.reduce((acc, data) => ({ ...acc, ...data }), {} as T);
+
+export type IPollRequestPromise<T> = Promise<Record<string, T>> & {
+  requests: Record<string, Promise<T>>;
+};
+
+export const mergeAllPollRequestPromises = <T extends object>(
+  results: Array<IPollRequestPromise<T>>,
+): IPollRequestPromise<T> => {
+  return Object.assign(Promise.all(results).then(mergeAll), {
```

Comment:
What does the `Object.assign` do on a promise?

Reply:
its like this
```TypeScript
const promise = Promise.all(results).then(mergeAll)
promise.requests = // the requests promises
```
it lets the user to await for the final result by using
```TypeScript
const finalResult = await promise;
```
or await for each individual request
```TypeScript
await promise.requests["request-key-one"]
```
We have it because poll is a batch request. This means you can await for the entire result to be ready or just await for a specific request.